From 859a205e307e99eab0c7ba56175cdb1e3c7c184b Mon Sep 17 00:00:00 2001
From: Davide Gerhard <rainbow@irh.it>
Date: Tue, 14 Apr 2020 10:01:33 +0200
Subject: [PATCH] gnuradio-companion: add Meta modifier key for macOS

implements support for the Meta key on macOS.
Others platforms continue to use Ctrl.
---
 grc/gui/Actions.py |  6 +++++-
 grc/gui/Dialogs.py |  4 +++-
 grc/gui/Utils.py   | 22 ++++++++++++++++++++++
 3 files changed, 30 insertions(+), 2 deletions(-)

diff --git a/grc/gui/Actions.py b/grc/gui/Actions.py
index cf81e63ac7..edbc4f8405 100644
--- grc/gui/Actions.py
+++ grc/gui/Actions.py
@@ -13,6 +13,7 @@
 
 from gi.repository import Gtk, Gdk, Gio, GLib, GObject
 
+from . import Utils
 
 log = logging.getLogger(__name__)
 
@@ -125,7 +126,10 @@ def __init__(self, name, parameter=None, label=None, tooltip=None,
         self.label = label
         self.tooltip = tooltip
         self.icon_name = icon_name
-        self.keypresses = keypresses
+        if keypresses:
+            self.keypresses = [kp.replace("<Ctrl>", Utils.get_modifier_key(True)) for kp in keypresses]
+        else:
+            self.keypresses = None
         self.prefix = prefix
         self.preference_name = preference_name
         self.default = default
diff --git a/grc/gui/Dialogs.py b/grc/gui/Dialogs.py
index af8e0230f1..f55d21bdb8 100644
--- grc/gui/Dialogs.py
+++ grc/gui/Dialogs.py
@@ -287,9 +287,10 @@ def show_help(parent):
         <u>Remove connection</u>: select the connection and press delete, or 
        drag the connection.
         \n\
-        *Press CTRL+K or see menu for Keyboard - Shortcuts
+        *Press Ctrl+K or see menu for Keyboard - Shortcuts
         \
     """)
+    markup = markup.replace("Ctrl", Utils.get_modifier_key())
 
     MessageDialogWrapper(
         parent, Gtk.MessageType.INFO, Gtk.ButtonsType.CLOSE, title='Help', markup=markup
@@ -328,6 +329,7 @@ def show_keyboard_shortcuts(parent):
             selected block.
     \
     """)
+    markup = markup.replace("Ctrl", Utils.get_modifier_key())
     
     MessageDialogWrapper(
         parent, Gtk.MessageType.INFO, Gtk.ButtonsType.CLOSE, title='Keyboard - Shortcuts', markup=markup
diff --git a/grc/gui/Utils.py b/grc/gui/Utils.py
index acfe054cfc..5d710532bd 100644
--- grc/gui/Utils.py
+++ grc/gui/Utils.py
@@ -8,6 +8,7 @@
 
 from __future__ import absolute_import
 
+from sys import platform
 import numbers
 
 from gi.repository import GLib
@@ -152,3 +153,24 @@ def scale(coor, reverse=False):
 def scale_scalar(coor, reverse=False):
     factor = Constants.DPI_SCALING if not reverse else 1 / Constants.DPI_SCALING
     return int(coor * factor)
+
+def get_modifier_key(angle_brackets=False):
+    """
+    Get the modifier key based on platform.
+
+    Args:
+        angle_brackets: if return the modifier key with <> or not
+
+    Returns:
+        return the string with the modifier key
+    """
+    if platform == "darwin":
+        if angle_brackets:
+            return "<Meta>"
+        else:
+            return "Meta"
+    else:
+        if angle_brackets:
+            return "<Ctrl>"
+        else:
+            return "Ctrl"
